src/librustc_mir/transform/fragment_locals.rs

```diff
@@ -0,0 +1,494 @@
+        let base_node = &mut self.locals[place.local];
+
+        // Avoid identity replacements, which would re-intern projections.
+        if let FragmentTreeNodeKind::OpaqueLeaf { replacement_local: None } = base_node.kind {
+            return None;
+        }
+
+        let (node, proj_elems) = base_node.project(place.projection);
+
+        Some(match node.kind {
+            FragmentTreeNodeKind::OpaqueLeaf { replacement_local } => Ok(Place {
+                local: replacement_local.expect("missing replacement"),
+                projection: self.tcx.intern_place_elems(proj_elems),
+            }),
+
+            // HACK(eddyb) this only exists to support `(Set)Discriminant` below.
+            FragmentTreeNodeKind::Nested(_) => {
+                assert_eq!(proj_elems, &[]);
+
+                Err(node)
+            }
+        })
+    }
+}
+
+impl MutVisitor<'tcx> for FragmentTreeReplacer<'tcx> {
+    fn tcx(&self) -> TyCtxt<'tcx> {
+        self.tcx
+    }
+
+    fn visit_place(&mut self, place: &mut Place<'tcx>, _: PlaceContext, _: Location) {
+        if let Some(place_replacement) = self.replace(place) {
+            match place_replacement {
+                Ok(place_replacement) => *place = place_replacement,
+                // HACK(eddyb) this only exists to support `(Set)Discriminant` below.
+                Err(_) => unreachable!(),
+            }
+        }
+    }
+
+    // Break up `VarDebugInfo` into fragments where necessary.
+    fn visit_var_debug_info(&mut self, var_debug_info: &mut VarDebugInfo<'tcx>) {
+        match &mut var_debug_info.contents {
+            VarDebugInfoContents::Compact(place) => {
+                if let Some(place_replacement) = self.replace(place) {
+                    match place_replacement {
+                        Ok(place_replacement) => *place = place_replacement,
+                        Err(node) => {
+                            let mut fragments = vec![];
+                            node.gather_debug_info_fragments(&mut vec![], &mut fragments);
+                            var_debug_info.contents =
+                                VarDebugInfoContents::Composite { ty: node.ty, fragments };
+                        }
+                    }
+                }
+            }
+            VarDebugInfoContents::Composite { ty: _, fragments } => {
+                for fragment in fragments {
+                    if let Some(place_replacement) = self.replace(&fragment.contents) {
+                        match place_replacement {
+                            Ok(place_replacement) => fragment.contents = place_replacement,
+                            // FIXME(eddyb) implement!!
+                            Err(_) => span_bug!(
+                                var_debug_info.source_info.span,
+                                "FIXME: implement fragmentation for {:?}",
+                                var_debug_info,
+                            ),
```

Comment:
I don't think this is possible to hit without MIR inlining (because only this pass can create `VarDebugInfoContents::Composite`, so it must've run in a callee that got inlined), but I need to implement it before landing anyway.
(`gather_debug_info_fragments` is already set up for this, I just forgot to do it)